Allos.Amazon.Sdk/S3/Transfer/Commands/MultipartUploadCommand.cs

```diff
@@ -39,15 +37,14 @@ internal class MultipartUploadCommand : BaseCommand
         /// <summary>
         /// Initializes a new instance of the <see cref="MultipartUploadCommand"/> class.
         /// </summary>
-        /// <param name="s3Client">The s3 client.</param>
-        /// <param name="config">The config object that has the number of threads to use.</param>
+        /// <param name="asyncTransferUtility">The <see cref="IAsyncTransferUtility"/> that created this command.</param>
         /// <param name="fileTransporterRequest">The file transporter request.</param>
         /// <param name="logger">Logger</param>
         internal MultipartUploadCommand(
-            IAmazonS3 s3Client, 
-            AsyncTransferConfig config, 
+            IAsyncTransferUtility asyncTransferUtility,
             UploadRequest fileTransporterRequest,
             ILogger logger)
+            : base(asyncTransferUtility, fileTransporterRequest)
```

Comment:
`UploadUnseekableStreamAsync` proposal:

```
protected virtual async Task UploadUnseekableStreamAsync(
    UploadRequest request, 
    CancellationToken cancellationToken = default)
{
    ArgumentNullException.ThrowIfNull(request.InputStream);
    
    int readBufferSize = S3Client.Config.BufferSize;
    long minPartSize = request.PartSize != 0 ? request.PartSize.ToInt64() : S3Constants.MinPartSize;

    RequestEventHandler requestEventHandler = (_, args) =>
    {
        if (args is WebServiceRequestEventArgs wsArgs)
        {
            string currentUserAgent = wsArgs.Headers[AWSSDKUtils.UserAgentHeader];
            wsArgs.Headers[AWSSDKUtils.UserAgentHeader] =
                currentUserAgent + " ft/s3-transfer md/UploadNonSeekableStream";
        }
    };

    var initiateRequest = ConstructInitiateMultipartUploadRequest(requestEventHandler);
    var initiateResponse = await S3Client.InitiateMultipartUploadAsync(initiateRequest, cancellationToken)
        .ConfigureAwait(false);

    try
    {
        var uploadPartResponses = new List<UploadPartResponse>();
        byte[] readBuffer = ArrayPool<byte>.Shared.Rent(readBufferSize);
        byte[] partBuffer = ArrayPool<byte>.Shared.Rent((int)minPartSize);

        try
        {
            await using var stream = request.InputStream;
            uint partNumber = 1;
            int bytesInPartBuffer = 0;
            int readBytesCount;

            do
            {
                readBytesCount = await stream.ReadAsync(readBuffer, 0, readBuffer.Length, cancellationToken)
                    .ConfigureAwait(false);

                if (readBytesCount > 0)
                {
                    if (bytesInPartBuffer + readBytesCount > partBuffer.Length)
                    {
                        // Ensure we only upload if partBuffer has data
                        if (bytesInPartBuffer > 0)
                        {
                            // The part buffer is full, upload it
                            await UploadPartBufferAsync(partBuffer, bytesInPartBuffer, partNumber, false, initiateResponse, uploadPartResponses, cancellationToken);
                            partNumber++;
                            bytesInPartBuffer = 0; // Reset the part buffer position
                        }
                    }

                    // Copy the read bytes into the part buffer
                    Array.Copy(readBuffer, 0, partBuffer, bytesInPartBuffer, readBytesCount);
                    bytesInPartBuffer += readBytesCount;
                }
            } while (readBytesCount > 0);

            // Upload any remaining data in the part buffer (this is the last part)
            if (bytesInPartBuffer > 0)
            {
                await UploadPartBufferAsync(partBuffer, bytesInPartBuffer, partNumber, true, initiateResponse, uploadPartResponses, cancellationToken);
            }
        }
        finally
        {
            ArrayPool<byte>.Shared.Return(partBuffer);
            ArrayPool<byte>.Shared.Return(readBuffer);
        }

        _uploadResponses = uploadPartResponses;
        CompleteMultipartUploadRequest compRequest = ConstructCompleteMultipartUploadRequest(
            initiateResponse, 
            true,
            requestEventHandler);
        await S3Client.CompleteMultipartUploadAsync(compRequest, cancellationToken).ConfigureAwait(false);
        Logger.Debug("Completed multi part upload (PartCount={PartCount}, UploadId={UploadId})",
            uploadPartResponses.Count, 
            initiateResponse.UploadId);
    }
    catch (Exception ex)
    {
        await S3Client.AbortMultipartUploadAsync(new AbortMultipartUploadRequest
        {
            BucketName = request.BucketName,
            Key = request.Key,
            UploadId = initiateResponse.UploadId
        }, cancellationToken).ConfigureAwait(false);
        Logger.Error(ex, "{ErrorMessage}", ex.Message);
        throw;
    }
}

private async Task UploadPartBufferAsync(
    byte[] buffer,
    int bytesInBuffer,
    uint partNumber,
    bool isLastPart,
    InitiateMultipartUploadResponse initiateResponse,
    List<UploadPartResponse> uploadPartResponses,
    CancellationToken cancellationToken)
{
    using var partStream = new MemoryStream(buffer, 0, bytesInBuffer, writable: false);
    var uploadPartRequest = ConstructUploadPartRequestForNonSeekableStream(
        partStream,
        partNumber,
        (ulong)bytesInBuffer,
        isLastPart,
        initiateResponse);

    var partResponse = await S3Client.UploadPartAsync(uploadPartRequest, cancellationToken).ConfigureAwait(false);
    Logger.Debug(
        "Uploaded part {PartNumber} (PartSize={PartSize} UploadId={UploadId} IsLastPart={IsLastPart})", 
        partNumber, 
        bytesInBuffer, 
        initiateResponse.UploadId,
        isLastPart);
    uploadPartResponses.Add(partResponse);
}
```

Reply:
@KirillTonic this one breaks unit tests;  I did not investigate but may return to it later